File: 78. Subsets-2.cpp
class Solution {
public:
  vector<vector<int>> res;
  vector<int> subset;
  vector<vector<int>> subsets(vector<int> &nums) {
    helper(nums, 0);

    return res;
  }

  void helper(vector<int> &nums, int start) {
    res.push_back(subset);

    for (int i = start; i < nums.size(); ++i) {
      subset.push_back(nums[i]);
      helper(nums, i + 1);
      subset.pop_back();
    }
  }
};

File: 78. Subsets-2.py
class Solution:
    def subsets(self, nums: List[int]) -> List[List[int]]:
        res = []
        path = []
        n = len(nums)

        def dfs(i):
            res.append(path.copy())
            for j in range(i, n):
                path.append(nums[j])
                dfs(j + 1)
                path.pop()

        dfs(0)
        return res
